File: telegram-bot/src/convert.js
const currency = require('./currency');
const db = require('./db');
const { code, formatNumber } = require('./format');
const { convertKeyboard } = require('./keyboards');

// Computes conversion of `amount` in `baseCode` into every one of the
// user's preferred currencies. Returns null if the user has no preferences.
async function computeConversion(userId, amount, baseCode, { force = false } = {}) {
  const preferred = db.getPreferences(userId);
  if (preferred.length === 0) return null;

  const base = baseCode.toLowerCase();
  const { date, rates } = await currency.getRates(base, { force });

  const results = [];
  for (const targetCode of preferred) {
    let rate;
    if (targetCode === base) {
      rate = 1;
    } else if (rates[targetCode] !== undefined) {
      rate = rates[targetCode];
    } else {
      continue; // currency not present for this base (e.g. delisted), skip quietly
    }
    const value = amount * rate;
    results.push({ code: targetCode, value, formatted: formatNumber(value) });
  }

  return { date, base, amount, results };
}

function buildConversionMessage(conversion) {
  const { date, base, amount, results } = conversion;
  const amountStr = formatNumber(amount);
  const header = `*Converting* ${code(`${amountStr} ${base.toUpperCase()}`)}`;

  if (results.length === 0) {
    return `${header}\n\nNone of your preferred currencies have rate data for this base right now\\.`;
  }

  const lines = results.map((r) => `*${r.code.toUpperCase()}* — ${code(r.formatted)}`);
  return `${header}\n\n${lines.join('\n')}\n\n_Rates as of ${code(date)}_`;
}

// Sends a fresh conversion result as a new chat message, backed by an
// `interactions` row so the Refresh/Copy buttons keep working after restarts.
async function sendConversion(ctx, userId, amount, baseCode) {
  const conversion = await computeConversion(userId, amount, baseCode);
  if (!conversion) {
    return ctx.replyWithMarkdownV2('You have not picked any preferred currencies yet\\. Use /setpreferred first\\.');
  }

  const interactionId = db.createInteraction('convert', userId, { amount, base: baseCode.toLowerCase() });
  const text = buildConversionMessage(conversion);
  return ctx.replyWithMarkdownV2(text, convertKeyboard(interactionId, conversion.results, true));
}

module.exports = { computeConversion, buildConversionMessage, sendConversion };

File: telegram-bot/src/convert.py
from . import currency, db
from .format import escape_md, format_number, table
from .keyboards import convert_keyboard
from .reply import send_rich_message


# Computes conversion of `amount` in `base_code` into every one of the
# user's preferred currencies. Returns None if the user has no preferences.
async def compute_conversion(user_id, amount, base_code, force=False):
    preferred = db.get_preferences(user_id)
    if not preferred:
        return None

    base = base_code.lower()
    result = await currency.get_rates(base, force=force)
    date, rates = result['date'], result['rates']

    results = []
    for target_code in preferred:
        if target_code == base:
            rate = 1
        elif target_code in rates:
            rate = rates[target_code]
        else:
            continue  # currency not present for this base (e.g. delisted), skip quietly
        value = amount * rate
        results.append({'code': target_code, 'value': value, 'formatted': format_number(value)})

    return {'date': date, 'base': base, 'amount': amount, 'results': results}


# Builds the conversion result as a Rich Message: a heading, a table of
# preferred currency -> converted amount, and the rate date. The fallback
# carries the same information as plain text.
def build_conversion_view(conversion, interaction_id, include_copy_buttons):
    date, base, amount, results = (
        conversion['date'],
        conversion['base'],
        conversion['amount'],
        conversion['results'],
    )
    title = f'Converting {format_number(amount)} {base.upper()}'

    if not results:
        note = 'None of your preferred currencies have rate data for this base right now.'
        rich = {
            'markdown': f'# {escape_md(title)}\n\n{note}',
            'fallback': f'{title}\n\n{note}',
        }
    else:
        rich = {
            'markdown': (
                f'# {escape_md(title)}\n\n'
                + table(['Amount'], [(r['code'].upper(), r['formatted']) for r in results])
                + f'\n\n*Rates as of {escape_md(date)}*'
            ),
            'fallback': (
                f'{title}\n\n'
                + '\n'.join(f"{r['code'].upper()}: {r['formatted']}" for r in results)
                + f'\n\nRates as of {date}'
            ),
        }

    buttons = convert_keyboard(interaction_id, results, include_copy_buttons)
    return rich, buttons


# Sends a fresh conversion result as a new chat message, backed by an
# `interactions` row so the Refresh/Copy buttons keep working after restarts.
async def send_conversion(event, user_id, amount, base_code):
    conversion = await compute_conversion(user_id, amount, base_code)
    if not conversion:
        return await event.respond(
            'You have not picked any preferred currencies yet. Use /setpreferred first.',
            parse_mode='md',
        )

    interaction_id = db.create_interaction('convert', user_id, {'amount': amount, 'base': base_code.lower()})
    rich, buttons = build_conversion_view(conversion, interaction_id, True)
    return await send_rich_message(event.client, event.chat_id, rich, buttons)
